homeassistant/components/google/calendar.py

```diff
@@ -138,15 +149,28 @@ async def async_setup_entry(
                     "has been imported to the UI, and should now be removed from google_calendars.yaml"
                 )
             entity_name = data[CONF_DEVICE_ID]
+            unique_id = f"{config_entry.unique_id}-{calendar_id}-{entity_name}"
```

Comment:
I'm not sure we should add the `entity_name` to the unique_id. At least not for config entries that don't have any YAML calendar config stored. The `entity_name` is generated from the calendar summary which could change. We don't want the unique id to change.

For config entries without YAML calendar config, each calendar can only represent a single entity so using the calendar id for unique id should be enough as suffix.

Reply:
Maybe remove the unique id for legacy entities that come from the YAML calendar config?

Reply:
OK, updated. My impression was that entries without a unique id do not have entity registry entries, so i've removed them from the entity registry entry. Let me know if that is not right.

I'm treating the YAML + single entity case still the same as if there was no yaml -- the default with no intervention by the user -- which is sightly different than what you said.

Aside: I'm thinking it may be time to take even more aggressive steps away from yaml, so maybe time to start discussing that. My best idea was to make an entity helper (e.g. kind of like a template binary sensor based on calendar trigger?). One question is how exactly to structure that, and the other is how to best match the server side search behavior (plus a few other behavior changes). Wisdom/thoughts here would be appreciated.

Reply:
Do we need anything to replace the YAML entities? Isn't the trigger enough?

Reply:
The primary interesting use case for YAML today is for search, display of next event, and more advanced conditions.  Users use this in two ways that are interesting to discuss here:

-  For displaying the next upcoming event matching some string.  This can be replaced with a template binary sensor that does it's own string matching based on the summary or description
- For triggering an automation with more advanced conditions. Calendar trigger can do a lot, but also the binary sensor holding state is useful for more advanced conditions, for example [lights and sunlight](https://community.home-assistant.io/t/evening-light-schedule-driven-by-google-calendar/384550) where the the current trigger may not contain the needed state

My impression is if we have `google_calendars.yaml` like this:

```yaml
- cal_id: example@group.calendar.google.com.  # For my "Device Automations" calendar
  entities:
  - device_id: backyard_light
    name: Backyard Light
    search: "#Backyard"
```

Then users need a binary sensory template to power those above two use cases, something like this maybe:

```yaml
template:
  - trigger:
      - platform: calendar
        event: start
        entity_id: calendar.device_automation
      - platform: calendar
        event: end
        entity_id: calendar.device_automation
    binary_sensor:
      - name: backyard_light
        state: {{ if('#Backyard' in trigger.calendar_event.summary, if(trigger.event == "start", "on", "off"), states("binary_sensor.front_light_schedule") }}
```

There are a few other smaller details:
- Who knows exactly how server side search works, e.g. multiple terms, etc so it may be hard to migrate
- May need to handle other attributes in addition to state
- Users that use the `!!` offset syntax that will need to manually migrate.  We no longer support dynamic offsets based on the summary information. I claim that can't really be supported well anyway given the corner cases around multiple events.

I think lots of users really like the search functionality and using these calendar entity for display, and i'm not yet seeing the simple migration path for them. I'm generally wondering if more love for complex calendar automations, or helpers, etc are needed first.

Reply:
I previously had a documentation PR to include this type of binary sensor template detail in our calendar automation page instructions, but we decided it was a bit too low level.

Reply:
I think this is a topic for the architecture repo. There may be more calendar integrations than google that have use cases for searching for events. It's also too complex to find a quick solution for in this PR.

Reply:
Thanks, i'll do a little more research then write up something for discussion with a few paths to explore.